R-packages/evalcast/R/get_predictions.R

```diff
@@ -69,22 +83,36 @@ get_predictions <- function(forecaster,
                             apply_corrections = function(signals) signals,
                             response_data_source = signals$data_source[1],
                             response_data_signal = signals$signal[1],
-                            forecaster_args = list()) {
-  
+                            parallel_execution = TRUE,
+                            forecaster_args = list(),
+                            honest_as_of = TRUE,
+                            offline_signal_dir = NULL) {
+
   assert_that(is_tibble(signals), msg="`signals` should be a tibble.")
-  
-  
-  out <- forecast_dates %>%
-    map(~ do.call(
-          get_predictions_single_date,
-          list(forecaster = forecaster,
-               signals = signals,
-               forecast_date = .x,
-               apply_corrections = apply_corrections,
-               forecaster_args = forecaster_args)
-          )
-        ) %>%
-    bind_rows()
+  assert_that(xor(honest_as_of, "as_of" %in% names(signals)), msg="`honest_as_of` should be set if and only if `as_of` is not set. Either remove as_of specification or set honest_as_of to FALSE.")
+
+  p = progressor(steps = length(forecast_dates))
+  get_predictions_single_date_ <- function(forecast_date) {
+    p()
+    do.call(get_predictions_single_date,
+            list(forecaster = forecaster,
+              signals = signals,
+              forecast_date = forecast_date,
+              apply_corrections = apply_corrections,
+              forecaster_args = forecaster_args,
+              honest_as_of = honest_as_of,
+              offline_signal_dir = offline_signal_dir))
+  }
+  if(parallel_execution) {
+    out <- mclapply(
```

Comment:
I'm going to remove the progress bar. It's not critical.

Reply:
However, even with the bar present, I wasn't able to reproduce your error...

```
r$> evalcast::get_predictions( 
        forecaster = evalcast::baseline_forecaster, 
        name_of_forecaster = "baseline", 
        signals = forecaster_signals[["baseline1_forecaster"]], 
        forecast_dates = seq(as.Date("2020-12-01"), as.Date("2020-12-31"), by=7), 
        incidence_period = "epiweek", 
        parallel_execution = FALSE, 
        forecaster_args = list(ahead = 3) 
    )                                                                                                                                                                                                                                                                      
Downloading confirmed_admissions_covid_1d from covidcast for period from 2020-08-01 to 2020-12-01.
Downloading confirmed_admissions_covid_1d from covidcast for period from 2020-08-01 to 2020-12-08.
Downloading confirmed_admissions_covid_1d from covidcast for period from 2020-08-01 to 2020-12-15.
Downloading confirmed_admissions_covid_1d from covidcast for period from 2020-08-01 to 2020-12-22.
Downloading confirmed_admissions_covid_1d from covidcast for period from 2020-08-01 to 2020-12-29.
...
Warning messages:
1: Data not fetched for the following days: 2020-11-29, 2020-11-30, 2020-12-01 
2: Data not fetched for the following days: 2020-12-06, 2020-12-07, 2020-12-08 
3: Fetching confirmed_admissions_covid_1d from hhs for 2020-12-13 to 2020-12-15 in geography '*': no results 
4: Data not fetched for the following days: 2020-12-21, 2020-12-22 
5: Data not fetched for the following days: 2020-12-28, 2020-12-29 

r$> evalcast::get_predictions( 
        forecaster = evalcast::baseline_forecaster, 
        name_of_forecaster = "baseline", 
        signals = forecaster_signals[["baseline1_forecaster"]], 
        forecast_dates = seq(as.Date("2020-12-01"), as.Date("2020-12-31"), by=7), 
        incidence_period = "epiweek", 
        parallel_execution = TRUE, 
        forecaster_args = list(ahead = 3) 
    )                                                                                                                                                                                                                                                                      
Downloading confirmed_admissions_covid_1d from covidcast for period from 2020-08-01 to 2020-12-01.
Downloading confirmed_admissions_covid_1d from covidcast for period from 2020-08-01 to 2020-12-08.
Downloading confirmed_admissions_covid_1d from covidcast for period from 2020-08-01 to 2020-12-15.
Downloading confirmed_admissions_covid_1d from covidcast for period from 2020-08-01 to 2020-12-22.
Downloading confirmed_admissions_covid_1d from covidcast for period from 2020-08-01 to 2020-12-29.
...
```

Reply:
Added the following to catch NULL results

```
get_predictions_single_date_ <- function(forecast_date) {
    preds <- do.call(get_predictions_single_date,
      list(
        forecaster = forecaster,
        signals = signals,
        forecast_date = forecast_date,
        apply_corrections = apply_corrections,
        forecaster_args = forecaster_args,
        honest_as_of = honest_as_of,
        offline_signal_dir = offline_signal_dir
      )
    )
    if (is.null(preds)) abort("A child process was terminated and returned a null .")
    return(preds)
  }
```

Though, according to [the docs](https://search.r-project.org/CRAN/refmans/bettermc/html/mclapply.html), a fatal child event will make mclapply fail by default.

Reply:
I don't think this would catch the NULLs, as it is inside the function run by mclapply.  The issue I've had is that NULLs appear within the list returned by mclapply, when the function would never return NULL regularly, due to the child process being killed for memory.

The docs you linked are for `bettermc`, not `parallel`, which looks like it's designed to fix the problem!  Are we actually set up to use this package right now?

Reply:
Oh jeez, I didn't even notice I linked to a wrapper for `parallel`. We can probably just drop it in, since we don't depend on complicated uses of mclapply.